src/oidcc_token.erl

```diff
@@ -814,15 +820,65 @@ validate_id_token(IdToken, ClientContext, Nonce) ->
         end
     end.
 
+%% @doc Authorization headers
+%%
+%%   Generate a map of authorization headers to use when using the given
+%%   access token to access an API endpoint.
+%%
+%% <h2>Examples</h2>
+%%
+%% ```
+%% {ok, ClientContext} =
+%%   oidcc_client_context:from_configuration_worker(provider_name,
+%%                                                  <<"client_id">>,
+%%                                                  <<"client_secret">>),
+%%
+%% %% Get Access Token record from somewhere
+%%
+%% Headers =
+%%   oidcc:authorization_headers(AccessTokenRecord, :get, Url, ClientContext).
+%% '''
+%% @end
+%% @since 3.2.0
+-spec authorization_headers(AccessTokenRecord, Method, Endpoint, ClientContext) -> HeaderMap when
+    AccessTokenRecord :: access(),
+    Method :: post | get,
+    Endpoint :: uri_string:uri_string(),
+    ClientContext :: oidcc_client_context:t(),
+    HeaderMap :: #{binary() => binary()}.
+-spec authorization_headers(AccessTokenRecord, Method, Endpoint, ClientContext, Opts) ->
+    HeaderMap
+when
+    AccessTokenRecord :: access(),
+    Method :: post | get,
+    Endpoint :: uri_string:uri_string(),
+    ClientContext :: oidcc_client_context:t(),
+    Opts :: #{dpop_nonce => binary()},
+    HeaderMap :: #{binary() => binary()}.
+authorization_headers(AccessTokenRecord, Method, Endpoint, ClientContext) ->
+    authorization_headers(AccessTokenRecord, Method, Endpoint, ClientContext, #{}).
+
+authorization_headers(
+    #oidcc_token_access{} = AccessTokenRecord,
+    Method,
+    Endpoint,
+    #oidcc_client_context{} = ClientContext,
+    Opts
+) ->
+    #oidcc_token_access{token = AccessToken, type = AccessTokenType} = AccessTokenRecord,
+    Header = oidcc_auth_util:add_authorization_header(
+        AccessToken, AccessTokenType, Method, Endpoint, Opts, ClientContext
+    ),
+    maps:from_list([{list_to_binary(Key), list_to_binary([Value])} || {Key, Value} <- Header]).
+
 -spec verify_aud_claim(Claims, ClientId) -> ok | {error, error()} when
     Claims :: oidcc_jwt_util:claims(), ClientId :: binary().
-verify_aud_claim(#{<<"aud">> := Audience} = Claims, ClientId) when is_list(Audience) ->
-    case lists:member(ClientId, Audience) of
-        true -> ok;
-        false -> {error, {missing_claim, {<<"aud">>, ClientId}, Claims}}
+verify_aud_claim(#{<<"aud">> := Audience} = Claims, ClientId) ->
+    case Audience of
+        ClientId -> ok;
+        [ClientId] -> ok;
+        _ -> {error, {missing_claim, {<<"aud">>, ClientId}, Claims}}
     end;
-verify_aud_claim(#{<<"aud">> := ClientId}, ClientId) ->
-    ok;
```

Comment:
Why are you doing this change? If the audiences are a list, the client id can be one of many entries.

Reply:
This came out of an FAPI2 conformance suite failure, in `fapi2-security-profile-id2-client-test-invalid-secondary-aud`:

> This test issues an id_token where the 'aud' is an array which contains both the correct client_id and the id for a non-existent client. This test should end with the client displaying an error message that there is an untrusted aud value in the id_token from the token_endpoint, or that the 'azp' claim is missing.
> 
> As per OpenID Connect section 3.1.3.7 clause 3:
>
> 'The ID Token MUST be rejected if the ID Token does not list the Client as a valid audience, or if it contains additional audiences not trusted by the Client.'
>
> and clause 4:
>
> 'If the ID Token contains multiple audiences, the Client SHOULD verify that an azp Claim is present.'

Reply:
I believe we then have to add extra checks for the other audiences in the list and not require a list that only contains the client id.

The application I'm working on right now at @Sustema-AG (using @zitadel) for example contains multiple audiences. One for the client but also one for the "project" the client is defined in.

Reply:
I'll update the implementation to check for an `azp` claim in this situation as well.

Reply:
Also we have to consider the `SHOULD` and not `MUST` in this spec:

> 'If the ID Token contains multiple audiences, the Client SHOULD verify that an azp Claim is present.'

I've seen multiple providers (for example Auth0 and Zitadel) that recommend using a token retrieved by an SPA for multiple APIs. Since `azp` can only contain one, it is not included.

We'll probably need some kind of security level config for FAPI / other modes that allows people to get away without an `azp` when having multiple `aud`.

There's also other extra requirements required by FAPI (see https://github.com/erlef/oidcc/discussions/267) that would need some kind of switch to enable a more secure mode.

Reply:
Yep, 💯 agree on the other requirements; I had some thoughts about it that I'll write up later. 

For now, I'll revert this change.

Reply:
Opened https://github.com/erlef/oidcc/discussions/316 to discuss implementing profile options.